### PR TITLE
Simplification to the keyword completion provider code.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs [hunk 1]
```diff
@@ -37,21 +34,21 @@ protected virtual Task<bool> IsValidContextAsync(int position, CSharpSyntaxConte
...
             {
-                return SpecializedCollections.SingletonEnumerable(
+                return ImmutableArray.Create(
                     new RecommendedKeyword(SyntaxFacts.GetText(syntaxKind.Value),
                         shouldFormatOnCommit: ShouldFormatOnCommit,
                         matchPriority: ShouldPreselect(context, cancellationToken) ? SymbolMatchPriority.Keyword : MatchPriority.Default));
             }
 
-            return null;
+            return ImmutableArray<RecommendedKeyword>.Empty;
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs:45]: ⚠️ This path is now allocating a `Task<TResult>`, where it wasn't before. Is it called often?

File: src/Features/Core/Portable/Completion/Providers/IKeywordRecommender.cs [hunk 2]
```diff
@@ -2,16 +2,13 @@
...
+using System.Collections.Immutable;
 using System.Threading;
-using System.Threading.Tasks;
 
 namespace Microsoft.CodeAnalysis.Completion.Providers
 {
     internal interface IKeywordRecommender<TContext>
     {
-        Task<IEnumerable<RecommendedKeyword>> RecommendKeywordsAsync(int position, TContext context, CancellationToken cancellationToken);
+        ImmutableArray<RecommendedKeyword> RecommendKeywords(int position, TContext context, CancellationToken cancellationToken);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/Providers/IKeywordRecommender.cs:12]: made immutable and non-async.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs [hunk 3]
```diff
@@ -84,17 +81,16 @@ internal TestAccessor GetTestAccessor()
...
                 if (syntaxKind.HasValue)
                 {
                     var matchPriority = _recommender.ShouldPreselect(context, CancellationToken.None) ? SymbolMatchPriority.Keyword : MatchPriority.Default;
-                    return SpecializedCollections.SingletonEnumerable(
-                        new RecommendedKeyword(SyntaxFacts.GetText(syntaxKind.Value), matchPriority: matchPriority));
+                    return ImmutableArray.Create(new RecommendedKeyword(SyntaxFacts.GetText(syntaxKind.Value), matchPriority: matchPriority));
                 }
 
-                return null;
+                return ImmutableArray<RecommendedKeyword>.Empty;
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs:82]: ⚠️ This path is now allocating a Task<TResult>, where it wasn't before. Is it called often?

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/DynamicKeywordRecommender.cs [hunk 4]
```diff
@@ -27,14 +28,11 @@ private static bool IsValidContext(int position, CSharpSyntaxContext context, Ca
...
+        public Task<ImmutableArray<RecommendedKeyword>> RecommendKeywordsAsync(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
         {
-            if (IsValidContext(position, context, cancellationToken))
-            {
-                return Task.FromResult(SpecializedCollections.SingletonEnumerable(new RecommendedKeyword("dynamic")));
-            }
-
-            return SpecializedTasks.Null<IEnumerable<RecommendedKeyword>>();
+            return IsValidContext(position, context, cancellationToken)
+                ? Task.FromResult(ImmutableArray.Create(new RecommendedKeyword("dynamic")))
```
Comment (sharwell) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/DynamicKeywordRecommender.cs]: 💡 Assuming that `RecommendedKeyword` is immutable, this result could be cached in a `static readonly` field

File: src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs [hunk 5]
```diff
@@ -56,58 +43,57 @@ public override async Task ProvideCompletionsAsync(CompletionContext context)
...
-                return null;
-            }
+                return ImmutableArray<RecommendedKeyword>.Empty;
 
-            var set = new HashSet<RecommendedKeyword>();
+            using var _1 = ArrayBuilder<Task<ImmutableArray<RecommendedKeyword>>>.GetInstance(out var tasks);
             foreach (var recommender in _keywordRecommenders)
+                tasks.Add(Task.Run(() => recommender.RecommendKeywordsAsync(position, context, cancellationToken), cancellationToken));
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs:74]: ❔ Do we have any indication that sequential execution here is problematic? I would expect most calls to `RecommendKeywordsAsync` to complete synchronously, which would make the `Task.WhenAll` unnecessary and potentially expensive.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs:45]: This no longer allocates tasks at all.

--

nearly all the allocations in keyword completin can likely be entirely removed.  i'm considering doing that, but it's a lot more mechanical work.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/DynamicKeywordRecommender.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/Completion/KeywordRecommenders/DynamicKeywordRecommender.cs]: yes.  it *definitely* could.  however, that's much more manual.  something i'd like to do later, but not hugely important right now.

File: src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs:74]: The goal would be to have the 100ish C# providers (or 200+ VB providers) just run as quickly as possible on all cores. I could do this as a parallel foreach i suppose, but this seemed appropriately idiomatic.

Comment (sharwell) [reply src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs:74]: ❗ Let's move the parallelism to a different PR. We can discuss it, but this PR is too big to deal with a change like that.

**Review by sharwell — CHANGES_REQUESTED** [review]
Need to remove the parallelism change from this refactoring PR.

File: src/EditorFeatures/CSharp.Wpf/Completion/CompletionProviders/CSharpReplCommandCompletionProvider.cs [hunk 6]
```diff
@@ -74,10 +74,9 @@ internal override bool IsInsertionTrigger(SourceText text, int characterPosition
 
         internal override ImmutableHashSet<char> TriggerCharacters { get; } = CompletionUtilities.SpaceTriggerCharacter;
 
-        protected override async Task<bool> ShouldDisplayCommandCompletionsAsync(SyntaxTree tree, int position, CancellationToken cancellationToken)
-        {
-            return await tree.IsBeforeFirstTokenAsync(position, cancellationToken).ConfigureAwait(false) &&
-                   tree.IsPreProcessorKeywordContext(position, cancellationToken);
-        }
+        protected override Task<bool> ShouldDisplayCommandCompletionsAsync(SyntaxTree tree, int position, CancellationToken cancellationToken)
+            => Task.FromResult(
```
Comment (sharwell) [inline src/EditorFeatures/CSharp.Wpf/Completion/CompletionProviders/CSharpReplCommandCompletionProvider.cs:78]: ⚠️  This needs to use cached return values, either manually from `SpecializedTasks` or automatically by using the state machine provided by `async`/`await`.

File: src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs [hunk 7]
```diff
@@ -14,317 +14,317 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Recommendations
     public class AbstractKeywordRecommenderTests : KeywordRecommenderTests
     {
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestAtRoot_Interactive()
+        public void TestAtRoot_Interactive()
         {
-            await VerifyKeywordAsync(SourceCodeKind.Script,
+            VerifyKeyword(SourceCodeKind.Script,
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs:19]: 📝 I'm assuming that all changes to the tests are:

* Return type change
* Removing `Async` suffix

And as such am not reviewing these changes in much detail. If there are any behavior changes please call them all out or move them to a different PR.

File: src/EditorFeatures/CSharpTest2/Recommendations/ByteKeywordRecommenderTests.cs [hunk 8]
```diff
@@ -680,37 +680,37 @@ static void Main(string[] args)
...
+            VerifyAbsence(@"
 /// <see cref=""System.$$"" />
 class C { }
 ");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestAfterAsync()
-            => await VerifyKeywordAsync(@"class c { async $$ }");
+        public void TestAfter()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/ByteKeywordRecommenderTests.cs:692]: 📝 Test name should not have changed

File: src/EditorFeatures/CSharpTest2/Recommendations/DecimalKeywordRecommenderTests.cs [hunk 9]
```diff
@@ -668,37 +668,37 @@ static void Main(string[] args)
...
+            VerifyAbsence(@"
 /// <see cref=""System.$$"" />
 class C { }
 ");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestAfterAsync()
-            => await VerifyKeywordAsync(@"class c { async $$ }");
+        public void TestAfter()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/DecimalKeywordRecommenderTests.cs:680]: 📝 Method name should not have changed

File: src/EditorFeatures/CSharpTest2/Recommendations/DoubleKeywordRecommenderTests.cs [hunk 10]
```diff
@@ -652,37 +652,37 @@ static void Main(string[] args)
...
+            VerifyAbsence(@"
 /// <see cref=""System.$$"" />
 class C { }
 ");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestAfterAsync()
-            => await VerifyKeywordAsync(@"class c { async $$ }");
+        public void TestAfter()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/DoubleKeywordRecommenderTests.cs:664]: 📝 Method name should not have changed

File: src/EditorFeatures/CSharpTest2/Recommendations/DynamicKeywordRecommenderTests.cs [hunk 11]
```diff
@@ -20,653 +20,653 @@ public class DynamicKeywordRecommenderTests : RecommenderTests
...
 @"class Program
 {
     #region $$
 }");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestAfterAsync()
-            => await VerifyKeywordAsync(@"class c { async $$ }");
+        public void TestAfter()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/DynamicKeywordRecommenderTests.cs:632]: 📝 Method name should not have changed

File: src/EditorFeatures/CSharpTest2/Recommendations/FloatKeywordRecommenderTests.cs [hunk 12]
```diff
@@ -625,37 +625,37 @@ static void Main(string[] args)
...
+            VerifyAbsence(@"
 /// <see cref=""System.$$"" />
 class C { }
 ");
         }
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestAfterAsync()
-            => await VerifyKeywordAsync(@"class c { async $$ }");
+        public void TestAfter()
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/FloatKeywordRecommenderTests.cs:637]: 📝 Method name should not have changed

File: src/EditorFeatures/CSharpTest2/Recommendations/ForEachKeywordRecommenderTests.cs [hunk 13]
```diff
@@ -13,190 +13,190 @@ namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Recommendations
     public class ForEachKeywordRecommenderTests : KeywordRecommenderTests
     {
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestAtRoot_Interactive()
+        public void TestAtRoot_Interactive()
         {
-            await VerifyKeywordAsync(SourceCodeKind.Script,
+            VerifyKeyword(SourceCodeKind.Script,
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest2/Recommendations/ForEachKeywordRecommenderTests.cs:18]: ❓ In the off chance we need to make this asynchronous in the future, can we keep the Verify methods asynchronous even though they currently run synchronously? It would save a lot of work both for this review and if something happens to change it again later.

File: src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Statements/CaseKeywordRecommender.vb [hunk 14]
```diff
@@ -48,7 +49,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.KeywordRecommenders.Stat
                 validKeywords.Add(New RecommendedKeyword("Case Else", VBFeaturesResources.Introduces_the_statements_to_run_if_none_of_the_previous_cases_in_the_Select_Case_statement_returns_True))
             End If
 
-            Return validKeywords
+            Return validKeywords.ToImmutableArray()
```
Comment (sharwell) [inline src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Statements/CaseKeywordRecommender.vb:52]: 💡  This is a great location for `TemporaryArray(Of RecommendedKeyword)` (the array always has 0-2 elements before calling `ToImmutableArray`)

File: src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Declarations/GetSetKeywordRecommender.vb [hunk 15]
```diff
@@ -71,7 +71,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Completion.KeywordRecommenders.Decl
                 recommendations.Add(New RecommendedKeyword("Set", VBFeaturesResources.Declares_a_Set_property_procedure_that_is_used_to_assign_a_value_to_a_property))
             End If
 
-            Return recommendations
+            Return recommendations.ToImmutableArray()
```
Comment (sharwell) [inline src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Declarations/GetSetKeywordRecommender.vb:74]: 💡 This is a good candidate for `TemporaryArray(Of RecommendedKeyword)` (always 0-2 elements)

**Comment by genlu** [comment]
>             var syntaxContextService = document.GetRequiredLanguageService<ISyntaxContextService>();

I assume the purpose of `ISyntaxContextService` is we will eventually move the cache from AbstractSymbolCompletionProvider to this service?

---
Refers to: src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs:52 in 7ab7168. [](commit_id = 7ab71683fe3ebb87411c1efe808fa9794eb0a081, deletion_comment = False)

**Comment by genlu** [comment]
>         private SyntaxKind? RecommendKeyword(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)

Rename this to `GetRecommendedKeyworkKind` or something like that

---
Refers to: src/Features/CSharp/Portable/Completion/KeywordRecommenders/AbstractSyntacticSingleKeywordRecommender.cs:50 in 7ab7168. [](commit_id = 7ab71683fe3ebb87411c1efe808fa9794eb0a081, deletion_comment = False)

File: src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Statements/CaseKeywordRecommender.vb (hunk 14 above)
Comment (CyrusNajmabadi) [reply src/Features/VisualBasic/Portable/Completion/KeywordRecommenders/Statements/CaseKeywordRecommender.vb:52]: agreed

**Comment by genlu** [comment]
>                 syntaxTree.IsBeforeFirstToken(position, cancellationToken);

Is this part of the change? I couldn't find the definition of `IsBeforeFirstToken`

---
Refers to: src/Features/CSharp/Portable/Completion/KeywordRecommenders/UndefKeywordRecommender.cs:23 in 7ab7168. [](commit_id = 7ab71683fe3ebb87411c1efe808fa9794eb0a081, deletion_comment = False)

File: src/EditorFeatures/CSharpTest2/Recommendations/ForEachKeywordRecommenderTests.cs (hunk 13 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest2/Recommendations/ForEachKeywordRecommenderTests.cs:18]: sure.

File: src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Completion/Providers/AbstractKeywordCompletionProvider.cs:74]: can do.

File: src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest2/Recommendations/AbstractKeywordRecommenderTests.cs:19]: yup.  this was all done mechanically too using our own refactoring tools.

**Comment by CyrusNajmabadi** [comment]
> Are you planning to eventually move the cache from AbstractSymbolCompletionProvider to this service, so it can be shared by all provider?

Yup :)

**Comment by CyrusNajmabadi** [comment]
> Is this part of the change? I couldn't find the definition of IsBeforeFirstToken

Yup.  

All our other syntax tree context helpers are synchronous and just do `tree.GetRoot(ct)`.  This just makes this single method follow that pattern, and allows this all to become synchronous.

File: src/EditorFeatures/CSharp.Wpf/Completion/CompletionProviders/CSharpReplCommandCompletionProvider.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharp.Wpf/Completion/CompletionProviders/CSharpReplCommandCompletionProvider.cs]: I made the entire codepath here synchronous without tasks.

**Comment by CyrusNajmabadi** [comment]
We're now caching many more keyword arrays as static fields to prevent allocations.

**Comment by CyrusNajmabadi** [comment]
@sharwell ptal I believe I got all your feedback addressed. Thanks!

File: src/EditorFeatures/CSharpTest2/Recommendations/ForEachKeywordRecommenderTests.cs (hunk 13 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest2/Recommendations/ForEachKeywordRecommenderTests.cs]: Thanks this was a huge help!

**Review by sharwell — APPROVED** [review]
